File: 백준/Gold/1461. 도서관/도서관.py
N, M = map(int, input().split())
locations = list(map(int, input().split()))
ans = 0
idx = -1
locations.sort();
for i in range(N):
    if locations[i] > 0:
        idx = i
        break
if idx != -1:
    minus = locations[:idx]
    plus = locations[idx:]
else:
    minus = locations[:]
    plus = []

minus.sort()
plus.sort(reverse=True)

for i in range(0, len(minus), M):
    tmp = abs(min(minus[i:i+M]))
    ans += tmp * 2

for i in range(0, len(plus), M):
    tmp = abs(max(plus[i:i+M]))
    ans += tmp * 2
if len(minus) > 0 and len(plus) > 0 :
    ans -= max(abs(minus[0]), abs(plus[0]))
elif len(minus) > 0:
    ans -= abs(minus[0])
elif len(plus) > 0:
    ans -= abs(plus[0])

print(ans)

File: 백준/Gold/1461. 도서관/도서관.java
import java.io.*;
import java.util.*;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        int[] books = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            int num = Integer.parseInt(st.nextToken());
            books[i] = num;
        }
        Arrays.sort(books);

        List<Integer> minus = new ArrayList<>();
        List<Integer> plus = new ArrayList<>();
        for (int i = 0; i < N; i++) {
            if (books[i] < 0) {
                minus.add(books[i]);
            } else {
                plus.add(books[i]);
            }
        }
        Collections.sort(plus, Collections.reverseOrder());

        int ans = 0;
        for (int i = 0; i < minus.size(); i += M) {
            int tmp = 0;
            for (int j = 0; j < M; j++) {
                tmp = Math.max(tmp, Math.abs(minus.get(i)));
            }
            ans += tmp * 2;
        }

        for (int i = 0; i < plus.size(); i += M) {
            int tmp = 0;
            for (int j = 0; j < M; j++) {
                tmp = Math.max(tmp, plus.get(i));
            }
            ans += tmp * 2;

        }

        if (plus.size() > 0 && minus.size() > 0 ) {
            ans -= Math.max(Math.abs(minus.get(0)), plus.get(0));
        } else if (plus.size() > 0) {
            ans -= plus.get(0);
        } else if (minus.size() > 0) {
            ans -= Math.abs(minus.get(0));
        }

        System.out.print(ans);
    }

}
